tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,201 @@
+    Vulkan = 2,
+    /// An OpenCL context
+    OpenCL = 3,
+    /// A D3D12 context.
+    Direct3D12 = 4,
+    /// A D3D11 context.
+    Direct3D11 = 5,
+}
+
+///
+#[derive(Clone)]
+pub struct GpuContext {
+    #[cfg(feature = "enable")]
+    client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+    #[cfg(feature = "enable")]
+    span_index: Arc<AtomicU16>,
+}
+#[cfg(feature = "enable")]
+static GPU_CONTEXT_INDEX: Lazy<Mutex<u8>> = Lazy::new(|| Mutex::new(0));
+
+///
+pub struct GpuSpan {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+    #[cfg(feature = "enable")]
+    // This is the index first query, the second query is always a wrapping_add(1) above this.
+    query_id: u16,
+    #[cfg(feature = "enable")]
+    ended: bool,
+}
+
+impl Client {
+    /// Creates a new GPU context.
+    ///
+    /// - 'name' is the name of the context.
+    /// - 'ty' is the type (backend) of the context.
+    /// - 'gpu_timestamp' is the gpu side timestamp the corresponds (as close as possible) to this call.
+    /// - 'period' is the period of the gpu clock in nanoseconds (setting 1.0 means the clock is 1GHz, 1000.0 means 1MHz, etc).
+    pub fn new_gpu_context(
+        self,
+        name: Option<&str>,
+        ty: GpuContextType,
+        gpu_timestamp: i64,
+        period: f32,
+    ) -> Option<GpuContext> {
+        #[cfg(feature = "enable")]
+        {
+            // We use a mutex to lock the context index to prevent races when using fetch_add.
+            let mut context_index_guard = GPU_CONTEXT_INDEX.lock().unwrap();
+            if *context_index_guard == 255 {
+                return None;
+            }
+            let context = *context_index_guard;
+            *context_index_guard += 1;
+            drop(context_index_guard);
+
+            unsafe {
+                sys::___tracy_emit_gpu_new_context(sys::___tracy_gpu_new_context_data {
+                    gpuTime: gpu_timestamp,
+                    period,
+                    context,
+                    flags: 0,
+                    type_: ty as u8,
+                })
+            };
+
+            if let Some(name) = name {
+                unsafe {
+                    sys::___tracy_emit_gpu_context_name_serial(
+                        sys::___tracy_gpu_context_name_data {
+                            context,
+                            name: name.as_ptr().cast(),
+                            len: name.len().try_into().unwrap_or(u16::MAX),
+                        },
+                    )
+                }
+            }
+
+            Some(GpuContext {
+                client: self,
+                context,
+                span_index: Arc::new(AtomicU16::new(0)),
+            })
+        }
+        #[cfg(not(feature = "enable"))]
+        Some(GpuContext {})
+    }
+}
+
+impl GpuContext {
+    ///
+    pub fn span_alloc(&self, name: &str, function: &str, file: &str, line: u32) -> GpuSpan {
```

Comment:
Could we also have a non-`_alloc` variant that takes `&'static SpanLocation`?

Reply:
Yes! Such an api didn't exist in the c api when I first wrote this MR, but it does now, so I'll add it.

Reply:
Welp. Turns out the function with the different name has exactly the same argument struct as the other one, so there's no difference in the calling code :|

Reply:
Based on the implementation of the analogous non-GPU API, I think the intention is that we reinterpret_cast the pointer into the same field:

https://github.com/wolfpld/tracy/blob/0e26b9e53fc88aeed75050c959af751e8c48828b/public/client/TracyProfiler.cpp#L3958C61-L3958C67

I didn't dig all the way down into the guts to see where this gets decoded, but there's a tag that tracy does indeed branch on to determine how to interpret the payload data.

Reply:
Nice hint! It works!